### PR TITLE
add associated error type to MessageBody

### PR DESCRIPTION
**Pull request by robjtede**
## PR Type

Feature

## PR Checklist

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt.
- [x] (Team) Label with affected crates and semver status.

## Overview
Step 1. Split out from the error remodelling, MessageBody was a large part of the pain because the error type was assumed to be `actix_http::Error` all the time.

### CODE REVIEW

File: actix-http/src/body/body.rs [hunk 1]
```diff
@@ -25,7 +26,7 @@ pub enum Body {
     Bytes(Bytes),
 
     /// Generic message body.
-    Message(Pin<Box<dyn MessageBody>>),
+    Message(BoxAnyBody),
```
Comment (robjtede) [inline actix-http/src/body/body.rs:29]: `Pin<Box<dyn MessageBody<Error = Error>>>` is the alternative and locks in that error type. This way, converting to and from StdError is shown to be viable.

File: actix-http/src/body/message_body.rs [hunk 2]
```diff
@@ -7,75 +7,99 @@ use std::{
...
     /// Body size hint.
     fn size(&self) -> BodySize;
 
     /// Attempt to pull out the next chunk of body bytes.
     fn poll_next(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-    ) -> Poll<Option<Result<Bytes, Error>>>;
-
-    downcast_get_type_id!();
```
Comment (robjtede) [inline actix-http/src/body/message_body.rs:26]: With some change to the macros (or inlining) it's possible to write `impl<E: 'static> dyn MessageBody<Error = E> + 'static` but i removed this at some point during writing this to fix an error.

I think it is possible to just use `std::any::Any` trait for any downcasting needs anyway (see altered test).

File: actix-http/src/service.rs [hunk 3]
```diff
@@ -59,6 +59,7 @@ where
     S::Response: Into<Response<B>> + 'static,
     <S::Service as Service<Request>>::Future: 'static,
     B: MessageBody + 'static,
+    B::Error: Into<Error>,
```
Comment (robjtede) [inline actix-http/src/service.rs:62]: having bounds look like this (for now) is a pretty well compatible change since trivially: `Error: Into<Error>`

**Comment by fakeshadow** [comment]
LGTM. 

What is wrong with concrete to `actix_http::Error` though. It's just a trait object that help generating a Response in dispatcher.

**Review by fakeshadow — APPROVED** [review]

**Comment by robjtede** [comment]
> What is wrong with concrete to actix_http::Error.

We need to move that type to -web along with `ResponseError` trait in order to make it so `error_response()` can return a `HttpResponse` instead of `Response`. Then introduce a new error type in -http that's not relying on `ResponseError` and mostly change these bounds to `B::Error: Into<Response<Body>>`.

It's an all or nothing situation tbh, if this isn't done then the `HttpResponse` changes will need to be reverted.

**Pull request merged**